### PR TITLE
make API much more flexible with different types of incoming requests

### PR DESCRIPTION
**Pull request by fiatjaf**
This is a dumb implementation I came up with. It seems to work whenever I test it (after I fixed various quirks and bugs), but there's probably some bug still. Anyway, the present request_wants_json, however, is not foolproof, it is failing and must be changed.

Most Javascript HTTP libraries, superagent, for example, automatically send the data you pass to them in JSON format and set the Content-Type header to application/json, but don't set the Accept header. And I can't imagine a situation in which someone would send an AJAX request like this and expect an HTML response.

Also, Angular automatically sets the Accept header to application/json, text/plain, */*, and I don't know why but Werkzeug parses it in a way that request.accept_mimetypes['application/json'] is 1, but request.accept_mimetypes['text/html'] is also 1, so it fails.

Besides this, I think we need to accept a JSON body when the user wants to send it, it will not hurt anyone or anybody to do it (in Javascript it is many times easier to send a JSON body, as the use cases above prove).

### CODE REVIEW

**Comment by fiatjaf** [comment]
I added a lot of tests (combinations of Accept and Content-Type) to see if Formspree would respond with the correct Content-Type. It seems to be working.

File: formspree/utils.py [hunk 1]
```diff
@@ -13,11 +13,27 @@
...
-        request.accept_mimetypes['text/html']
+    if accept_better('json', 'html'):
+        return True
+    if 'json' in request.headers.get('Content-Type', '') and \
+            not accept_better('html', 'json'):
+        return True
+
+
+def accept_better(subject, against):
+    if 'Accept' in request.headers:
```
Comment (cabello) [inline formspree/utils.py]: Please, return False in case `Accept` is not in request headers.

**Comment by colevscode** [comment]
Let's resolve merge conflicts and merge this

**Comment by fiatjaf** [comment]
@rohitdatta look at how strange is this diff: https://github.com/formspree/formspree/pull/67/files

This new PR I opened is just my local branch merged into a new branch which is itself branched from my local master (which was https://github.com/fiatjaf/formspree/commit/368a990b3c4583a87441e1c49c4e76ff8328f3eb, but should be https://github.com/fiatjaf/formspree/commit/3efa6e9e85cc9f95d50b768916bba68382a689fc and I have now updated my local repo and fiatjaf/formspree on GitHub) with two remaining conflicts fixed (I don't know why I didn't fix it in the previous commit).

I say we should be on https://github.com/fiatjaf/formspree/commit/3efa6e9e85cc9f95d50b768916bba68382a689fc because it is the last commit before merging the PR that resulted in so many problems.

If we were there, this would be the PR, without conflicts, that we should be merging now: https://github.com/fiatjaf/formspree/compare/master...fiatjaf:api-accepts-json-and-is-more-flexible

So I think I will force-push back to https://github.com/fiatjaf/formspree/commit/3efa6e9e85cc9f95d50b768916bba68382a689fc, then merge this new PR I opened, then we can move on.

Do you agree?

**Comment by rohitdatta** [comment]
So I'm always hesitant to force push in case it can mess something up, and I think if you're at 3efa6e9, we won't need to do it. I reset the changes I made in the branch `make-api-more-flexible`, so that it is now at the same commit as master. When I copied the changes in the 5 files on fiatjaf@fe03baf and ran tests, everything seemed to pass.

So I think the best course instead of force pushing would be maybe instead, just clone the Formspree repo directly for this, and then make your changes on that branch and push. Then we can make sure it passes the tests (which it will), and then you can merge. At that point, you can update your fork from the formspree master, and your fork will be at the same commit as the Formspree master repo.

I'm just hesitant to force push unless really necessary since I don't want the entire repository to get messed up just in case. If this doesn't work, then I think it might make sense to do it, but I'd like to try this first, since I'm pretty sure it should fix everything.

Is that ok with you?

**Comment by fiatjaf** [comment]
This is confusing. I don't think I get it. Should I merge my changes into `make-api-more-flexible`, then merge into `master`?

The problem is that I don't know what changes should I apply to master (or make-api-more-flexible), since it has changed so much since https://github.com/formspree/formspree/commit/3efa6e9e85cc9f95d50b768916bba68382a689fc.

Theoretically, if I could do what you're suggesting I should be able to merge https://github.com/formspree/formspree/pull/67 directly into master, but I don't think should even try. That would risk the codebase much more than a force-push.

Is that what you meant? Or didn't I understand it all?

**Comment by fiatjaf** [comment]
The change I'm suggesting is just a `git reset --hard 3efa6e9e85cc9f95d50b768916bba68382a689fc`, i.e., just go back in time, back to a time when we understood the codebase.

Then force-push, then merge #67. It will make us a clean and safe history.

But I'm happy with your approach also, I just don't understand it very well.

**Comment by rohitdatta** [comment]
So I reset the codebase back to 3efa6e9, it's just that it still has the intermediate commits, but on master it is exactly the same. I then took your commits and put them in #68. Travis CI says the tests are passing and when I diffed the files between that branch and the branch on your fork, it says they're exactly the same except for the README and one HTML file being out of date on your branch.

I think that should solve everything, but let me know if you wanted to do something different.

**Comment by fiatjaf** [comment]
Right, now that makes more sense. I'm fine. I'm sorry for the complaints. Please merge the PR.

**Comment by rohitdatta** [comment]
Alright sounds good! Sorry about messing it up to begin with.

**Pull request merged**